.github/workflows/publish.yml

```diff
@@ -0,0 +1,75 @@
+name: Publish
+
+on:
+  # push on main branch means to deploy production
+  push:
+    branches:
+      - main
+  # pull-request means to deploy preview
+  pull_request_target:
+  # workflow means to manuel deploy production
+  workflow_dispatch:
+  workflow_call:
+
+jobs:
+  build-and-upload:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    steps:
+      # we set the correct repository and reference sha for PR preview
+      - name: Checkout PR
+        if: ${{ github.event_name == 'pull_request_target' }}
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ github.event.pull_request.head.ref }}
+          repository: ${{ github.event.pull_request.head.repo.full_name }}
+
+      # we set the correct reference sha for production
+      - name: Checkout
+        if: ${{ github.event_name != 'pull_request_target' }}
+        uses: actions/checkout@v3
+        with:
+          ref: main
+
+      # extract pages setting
+      - name: Setup Pages
+        uses: actions/configure-pages@v1
+      # setup node environment
+      - name: Use Node.js
+        uses: actions/setup-node@v3
+        with:
+          node-version: 18
+          cache: npm
+          node-version-file: '.nvmrc'
+      # install dependencies
+      - name: Install Packages
+        run: npm install
+      # build website
+      - name: Build website
+        run: npm run build:website
+        env:
+          GH_TOKEN: ${{ github.token }}
+      - name: Upload artifact
+        uses: actions/upload-pages-artifact@v1
+        with:
+          path: './build'
+
+  deploy:
+    runs-on: ubuntu-latest
+    needs: build-and-upload
+    environment:
+      name: github-pages
+      url: ${{ steps.deployment.outputs.page_url }}
+    permissions:
+      pages: write
+      id-token: write
+    outputs:
+      deployment_url: ${{ steps.deployment.outputs.page_url }}
+    steps:
+      - name: Deploy to GitHub Pages
+        id: deployment
+        uses: actions/deploy-pages@v1
+        # enable preview when PR
+        with:
+          preview: ${{ github.event_name == 'pull_request_target' }}
```

Comment:
Should we try it 😈?

>    description: 'Is this attempting to deploy a pull request as a GitHub Pages preview site? (NOTE: This feature is only in alpha currently and is not available to the public!)'


I would keep the old workflow as `workflow_dispatch` only - just as backup

Reply:
> Should we try it 😈?

Yes, just to see if we are capable for this feature.